### PR TITLE
rec: log vm.max_map_count possibly being too low and log a few exceptions (with rate limiting)

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

- Warn when `vm.max_map_count` is dangerously low and adjust `max-mthreads` accordingly
- Fix a few cases where lack of logging could hide bad errors.
- Don't let recursorthead die on exceptions for the pdns-distributes-queries=no case
- Use rate limited logging for the logging in these cases

Not that the rate limited logging is not very smart, it throws all logging calls in one basket.

You might want to see part of the diff with whitespace ignored, it boils down to indenting changes due to an introduce try/catch block.

Fixes the logging part of #14180 

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
Do you think it makes sense to export the current number of mappings (number of lines in `/proc/$(pid)/maps` might work)?

**Comment by omoerbeek** [comment]
> Do you think it makes sense to export the current number of mappings (number of lines in `/proc/$(pid)/maps` might work)?

A bit hesitant, 

> Do you think it makes sense to export the current number of mappings (number of lines in `/proc/$(pid)/maps` might work)?

Might be useful indeed, though we should be careful, I know that in the past producing map info was very expensive (but maybe that was about `smap` ?

**Comment by omoerbeek** [comment]
Reworked the rate limiting code a bit.
While tetsing, I am seeing `FDMultiplexer::run() is not reentrant!`, which can hapen if the exception hapens in the `run()` method. In that case `d_inrun` is not reset.

**Review by rgacogne — CHANGES_REQUESTED** [review]
PR looks good, two suggestions

**Review by rgacogne — APPROVED** [review]

File: pdns/recursordist/rec-main.cc [hunk 1]
```diff
@@ -920,6 +923,27 @@ static void checkLinuxIPv6Limits([[maybe_unused]] Logr::log_t log)
...
+    auto lim = std::stoull(line);
+    // mthread stack use 3 maps per stack (2 guard pages + stack itself). Multiple by 4 for extra allowance.
+    // Also add 2 for handler and task threads.
+    auto workers = RecThreadInfo::numTCPWorkers() + RecThreadInfo::numUDPWorkers() + 2;
+    auto mapsNeeded = 4ULL * g_maxMThreads * workers;
+    if (lim < mapsNeeded) {
+      g_maxMThreads = static_cast<unsigned int>(lim / (4ULL * workers));
+      SLOG(g_log << Logger::Error << "sysctl vm.max_map_count= <" << mapsNeeded << ", this may cause 'bad_alloc' exceptions; adjusting max-mthreads to " << g_maxMThreads << endl,
+           log->info(Logr::Error, "sysctl vm.max_map_count < mapsNeeded, this may cause 'bad_alloc' exceptions, adjusting max-mthreads",
+                     "kern.max_map_count", Logging::Loggable(lim), "mapsNeeded", Logging::Loggable(mapsNeeded),
```
[the remark this answers is not in the document]
Comment (omoerbeek) [reply pdns/recursordist/rec-main.cc:940]: yes, thanks

**Pull request merged**